pkg/gadgets/dns/tracer/bpf/dns.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 The Inspektor Gadget authors */
+
+#include <linux/bpf.h>
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/in.h>
+#include <linux/udp.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+#include "dns-common.h"
+
+#define DNS_OFF (ETH_HLEN + sizeof(struct iphdr) + sizeof(struct udphdr))
+
+/* llvm builtin functions that eBPF C program may use to
+ * emit BPF_LD_ABS and BPF_LD_IND instructions
+ */
+unsigned long long load_byte(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.byte");
+unsigned long long load_half(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.half");
+unsigned long long load_word(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.word");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+} events SEC(".maps");
+
+// https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.1
+union dnsflags {
+	struct {
+		__u8 rd :1;	// recursion desired
+		__u8 tc :1;	// truncation
+		__u8 aa :1;	// authoritive answer
+		__u8 opcode :4;	// kind of query
+		__u8 qr :1;	// 0=query; 1=response
+
+		__u8 rcode :4;	// response code
+		__u8 z :3;	// reserved
+		__u8 ra :1;	// recursion available
+	};
+	__u16 flags;
+};
+
+struct dnshdr {
+	__u16 id;
+
+	union dnsflags flags;
+
+	__u16 qdcount; // number of question entries
+	__u16 ancount; // number of answer entries
+	__u16 nscount; // number of authority records
+	__u16 arcount; // number of additional records
+};
+
+SEC("socket1")
+int bpf_prog1(struct __sk_buff *skb)
+{
+	// Skip non-IP packets
+	if (load_half(skb, offsetof(struct ethhdr, h_proto)) != ETH_P_IP)
+		return 0;
+
+	// Skip non-UDP packets
+	if (load_byte(skb, ETH_HLEN + offsetof(struct iphdr, protocol)) != IPPROTO_UDP)
+		return 0;
+
+	// Skip non DNS Query packets
+	union dnsflags flags;
+	flags.flags = load_half(skb, DNS_OFF + offsetof(struct dnshdr, flags));
+
+	// Capture questions and ignore answers
+	if (flags.qr)
+		return 0;
+	// Capture recursive requests only
+	if (!flags.rd)
```

Comment:
Why do you want to ignore non-recursive queries? I agree that it would be quite strange that clients will not ask the DNS server to perform recursion for them but we have to consider it is not mandatory. In fact, it happened to me in the past (Notice it was in a completely different environment), that a client was using a customised DNS library where `RD` was not being set to `1` by default and the DNS server was not even answering the requests. In such a case, if you add this filter, we wouldn't be able to see those queries which could be misunderstood. Don't you think?

Reply:
It makes sense. I will remove this check.

Reply:
Done